### PR TITLE
Modify partition rewrite proxy listener to handler

### PR DESCRIPTION
**Pull request by dfangl**
## Motivation
Our partition rewrite listener, which was still implemented as ProxyListener in the old request handling framework, was more or less ineffective, as it would only work for services still being implemented in this framework.
To be able to use it properly again, and to get rid of this old framework at some point, a migration to a Handler is necessary.

## Changes
* Instead of a request and a response listener, the partition handler is now implemented as Handler
* The handler will proxy all requests back to the handler chain with requests being rewritten before / responses being rewritten afterwards. This allows us to have no impact on the rest of the chain, by not modifying werkzeug's requests objects.
* Tests are rewritten to make use of Request / Response objects
* Arn's without a region field (e.g. IAM) will be rewritten on response based on the region header set in the request. If it is not set, fallbacks as usual will commence.
* Changes in the handler are mostly `__call__`, `modify_request`, `modify_response`. All other changes are only for using the request region for arns without region.

## Limitations
* Streaming: Won't work for now. Maybe we can for now not proxy any streaming requests? Can confirm this kinesis subscription to shards is broken.
* Cross partition error handling: While cross partition requests are not allowed, we currently will not enforce anything like this.

### CODE REVIEW

File: localstack/aws/handlers/partition_rewriter.py [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+        forward_rewritten_headers["LS-INTERNAL-REWRITE-HANDLER"] = "1"
+        # Create a new request with the updated data
+        return Request(
+            method=request.method,
+            headers=forward_rewritten_headers,
+            path=forward_rewritten_path,
+            body=forward_rewritten_body,
+            # we have to set query string to None to avoid it being counted as defined in werkzeug
+            query_string=forward_rewritten_query_string,
+        )
```
Comment (thrau) [inline localstack/aws/handlers/partition_rewriter.py:106]: as you pointed out, doing the re-writing in this way is problematic because it blocks any streamed bodies. maybe it would be good to skip the arn rewriting for bodies that are streamed?

what you can do is to check 
```python
if context.operation.has_streaming_input:
    ...
```
and in that case, skip the re-writing, and to the new Request's body, pass `request._get_stream_for_parsing()` (i know that's not great, it's what we do in the parser:
https://github.com/localstack/localstack/blob/0aa9ffb17d6e0e76b0cab5f7f9ddca45bc46c685/localstack/aws/protocol/parser.py#L592-L597
https://github.com/localstack/localstack/blob/0aa9ffb17d6e0e76b0cab5f7f9ddca45bc46c685/localstack/aws/protocol/parser.py#L634-L644

basically this should solve the proxying issue, assuming that no other handler before that has consumed the stream. i'm not 100% sure this will work that way though tbh.

**Review by thrau — APPROVED** [review]
thanks for doing this daniel. this looks like a pretty tricky piece of code, and it looks very clean! i'm OK with merging as is, even given the blocking request limitation to get a fix out, but this shouldn't stay this way.

fundamentally i think we need to solve the streaming proxy issue (i.e., proxy streamed requests without consuming the request, but allowing us to seek to 0 if needed) for this to work properly. then we could theoretically solve the body re-writing by wrapping the stream iterator that underlies the request object. this is pretty involved though and should maybe be shepherded by @alexrashed or me. happy to do this in a follow up PR.

an alternative for now would be what i suggest in the comment, to basically skip body re-writing for streamed bodies. but not sure that will work well :shrug:
